cmd/snap-confine/udev-support.c

```diff
@@ -18,301 +18,390 @@
+		    devices_relpath, security_tag_relpath);
 	}
 
-	// nvidia modules are proprietary and therefore aren't in sysfs and
-	// can't be udev tagged. For now, just add existing nvidia devices to
-	// the cgroup unconditionally (AppArmor will still mediate the access).
-	// We'll want to rethink this if snapd needs to mediate access to other
-	// proprietary devices.
-	//
-	// Device major and minor numbers are described in (though nvidia-uvm
-	// currently isn't listed):
-	// https://github.com/torvalds/linux/blob/master/Documentation/admin-guide/devices.txt
-	char nv_path[15] = { 0 };	// /dev/nvidiaXXX
-	const char *nvctl_path = "/dev/nvidiactl";
-	const char *nvuvm_path = "/dev/nvidia-uvm";
-	const char *nvidia_modeset_path = "/dev/nvidia-modeset";
+	/* Open devices.allow relative to /sys/fs/cgroup/devices/snap.$SNAP_NAME.$APP_NAME */
+	const char *devices_allow_relpath = "devices.allow";
+	int devices_allow_fd = -1;
+	devices_allow_fd = openat(security_tag_fd, devices_allow_relpath,
+				  O_WRONLY | O_CLOEXEC | O_NOFOLLOW);
+	if (devices_allow_fd < 0) {
+		die("cannot open %s/%s/%s/%s", cgroup_path, devices_relpath,
+		    security_tag_relpath, devices_allow_relpath);
+	}
 
-	struct stat sbuf;
+	/* Open devices.deny relative to /sys/fs/cgroup/devices/snap.$SNAP_NAME.$APP_NAME */
+	const char *devices_deny_relpath = "devices.deny";
+	int devices_deny_fd = -1;
+	devices_deny_fd = openat(security_tag_fd, devices_deny_relpath,
+				 O_WRONLY | O_CLOEXEC | O_NOFOLLOW);
+	if (devices_deny_fd < 0) {
+		die("cannot open %s/%s/%s/%s", cgroup_path, devices_relpath,
+		    security_tag_relpath, devices_deny_relpath);
+	}
 
-	// /dev/nvidia0 through /dev/nvidia254
-	for (unsigned nv_minor = 0; nv_minor < 255; nv_minor++) {
-		sc_must_snprintf(nv_path, sizeof(nv_path), "/dev/nvidia%u",
-				 nv_minor);
+	/* Open cgroup.procs relative to /sys/fs/cgroup/devices/snap.$SNAP_NAME.$APP_NAME */
+	const char *cgroup_procs_relpath = "cgroup.procs";
+	int cgroup_procs_fd = -1;
+	cgroup_procs_fd = openat(security_tag_fd, cgroup_procs_relpath,
+				 O_WRONLY | O_CLOEXEC | O_NOFOLLOW);
+	if (cgroup_procs_fd < 0) {
+		die("cannot open %s/%s/%s/%s", cgroup_path, devices_relpath,
+		    security_tag_relpath, cgroup_procs_relpath);
+	}
 
-		// Stop trying to find devices after one is not found. In this
-		// manner, we'll add /dev/nvidia0 and /dev/nvidia1 but stop
-		// trying to find nvidia3 - nvidia254 if nvidia2 is not found.
-		if (stat(nv_path, &sbuf) != 0) {
-			break;
-		}
-		_run_snappy_app_dev_add_majmin(udev_s, nv_path,
-					       major(sbuf.st_rdev),
-					       minor(sbuf.st_rdev));
+	/* Everything worked so pack the result and "move" the descriptors over so
+	 * that they are not closed by the cleanup functions. */
+	fds.devices_allow_fd = devices_allow_fd;
+	fds.devices_deny_fd = devices_deny_fd;
+	fds.cgroup_procs_fd = cgroup_procs_fd;
+	devices_allow_fd = -1;
+	devices_deny_fd = -1;
+	cgroup_procs_fd = -1;
```

Comment:
Putting these back to -1 is fine but unneeded since they are scoped to this function and you are immediately returning.

Reply:
It is necessary. Please observe that each of the non-structure variables assigned above has a cleanup function so without resetting them to a safe value of -1, the return value would return stale descriptors because of the compiler-guaranteed call to `close()`

Reply:
From what I can see I concur with @zyga here - the cleanup function will run on scope exit (ie when this function returns) at which point it will `close()` these fds if they are not already -1 - in this case, we don't want them to get closed since they are now 'owned' by the caller within the fds struct so setting them to -1 ensures they remain open for the caller to close later.

Reply:
Perhaps that warrants a comment before `devices_allow_fd = -1` for the benefit of anyone who ends up wondering about this piece of code during future reviews/refactors.